### PR TITLE
8255380: (zipfs) ZipFileSystem::readExtra can fail if zipinfo-time is not set to false

### PR DESCRIPTION
**Pull request by LanceAndersen**
Hi,

Please review the fix for JDK-8255380 which addresses an issue when the Zip file is > 4GB.  Zip FS when processing the CEN extra data does not take into account the fact that there is no specific order to how the extra data fields are written.  Info-ZIP writes the fields in a different order than Zip FS which presents a problem when evaluating the Info-ZIP extended timestamp and the LOC offset is 0XFFFFFFFF therefore the LOC offset needs to be read from the EXTID_ZIP64 extra data prior to attempting to read the LOC extra data field.

The fix will defer reading of the LOC extra data field, if needed until all of the CEN extra data has been processed.

Using jdk.nio.zipfs.ZipInfo, you can see the ordering difference of the CEND extra data fields when using Zip FS and info-zip.

Info-zip is included with Mac OS so the test uses ProcessBuilder to execute zip on Mac OS and Linux.  

Mach5 tests jdk-tier1, jdk-tier2, and jdk-tier3 run cleanly.

Best,
Lance

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255380](https://bugs.openjdk.java.net/browse/JDK-8255380): (zipfs) ZipFileSystem::readExtra can fail if zipinfo-time is not set to false

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/987/head:pull/987`
`$ git checkout pull/987`

### CODE REVIEW

File: src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java [hunk 1]
```diff
@@ -3011,12 +2987,66 @@ private void readExtra(ZipFileSystem zipfs) throws IOException {
                 }
                 off += (sz + 4);
             }
+
+            // We need to read the LOC extra data and the LOC offset was obtained
+            // from the EXTID_ZIP64 field.
+            if(hasZip64LocOffset) {
```
Comment (cl4es) [inline src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java:2993]:
```suggestion
            if (hasZip64LocOffset) {
```

File: src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java [hunk 2]
```diff
@@ -2965,44 +2970,15 @@ private void readExtra(ZipFileSystem zipfs) throws IOException {
...
-                        if ((flag & 0x4) != 0 && locPos <= end) {
-                            ctime = unixToJavaTime(LG(buf, locPos));
-                        }
-                        break;
+                    // If the LOC offset is 0xFFFFFFFF, then we need to read the
+                    // LOC offset from the EXTID_ZIP64 extra data. Therefore
+                    // wait until all of the CEN extra data fields have been processed
+                    // prior to reading the LOC extra data field in order to obtain
+                    // the Info-ZIP Extended Timestamp.
+                    if(locoff != ZIP64_MINVAL) {
```
Comment (cl4es) [inline src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java:2978]:
```suggestion
                    if (locoff != ZIP64_MINVAL) {
```

**Review by cl4es — APPROVED** [review]
LGTM. Some nits inline.

I guess the test can be cause for issues in some test systems since it needs to write out the 4Gb+ file. Is this why you've only enabled it on linux and mac? Perhaps someone might have ideas on how to improve this.

**Comment by LanceAndersen** [comment]
> LGTM. Some nits inline.
> 
> I guess the test can be cause for issues in some test systems since it needs to write out the 4Gb+ file. Is this why you've only enabled it on linux and mac? Perhaps someone might have ideas on how to improve this.

To validate the test, requires info-zip which comes on Mac and linux.  It is not included with windows.  There are no issues if the Zip is created via java.util.zip or Zip FS

File: src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java (hunk 1 above)
Comment (LanceAndersen) [reply src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java]: Resolved

**Comment by cl4es** [comment]
> To validate the test, requires info-zip which comes on Mac and linux. It is not included with windows. There are no issues if the Zip is created via java.util.zip or Zip FS

Ah, gotcha. I see and understand the use of the external zip now. This seem a bit fragile - maybe needs to be more graceful in case the system zip doesn't exist or doesn't do what you expect?

**Comment by LanceAndersen** [comment]
> > To validate the test, requires info-zip which comes on Mac and linux. It is not included with windows. There are no issues if the Zip is created via java.util.zip or Zip FS
> 
> Ah, gotcha. I see and understand the use of the external zip now. This seem a bit fragile - maybe needs to be more graceful in case the system zip doesn't exist or doesn't do what you expect?

Zip is provided on Mac and Linux so unless the system is mis-configured, zip should be available on the path.  Are you suggesting to execute "zip -h" or "which zip" and  skip if zip is not there?  I could make it a manual test but thought given zip should be on linux and Mac that should be OK.  As long as the zip file is created, we are good to go.   I chose not to fall back to Zip FS (or java.util.zip)  to create the zip file as I know this will always pass.

Given the size of the file needed, it would be too large to check a file into the workspace.

Please let me know if you had another idea (Or prefer this to be a manual test)

**Comment by cl4es** [comment]
Given the need to generate a 4Gb file on the fly I do feel this maybe ought to be a manual test, or maybe there's some annotation to ensure it's not run on systems with too little disk space (and no compatible zip).

**Comment by LanceAndersen** [comment]
> Given the need to generate a 4Gb file on the fly I do feel this maybe ought to be a manual test, or maybe there's some annotation to ensure it's not run on systems with too little disk space (and no compatible zip).

OK,  I can make it a manual test, thank you for the feedback Claes!

**Review by cl4es — APPROVED** [review]

**Pull request closed without merging**